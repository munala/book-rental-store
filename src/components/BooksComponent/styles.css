.books-container {
  margin-top: 20px;
  margin-bottom: 50px;
  flex: 1;
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: stretch;
  width: 100%;
}

.books-list-container {
  padding: 0px 20px;
  background-color: #eaeaea;
  border-radius: 20px;
  overflow: scroll;
  padding: 0px;
  width: 100%;
}

.books-button {
  width: 164px;
}

.books-list {
  flex: 1;
  background-color: transparent;
  width: 100%;
}

.books-list:last-child {
  border-bottom-width: 0px;
}

.books-list-item {
  display: flex;
  background-color: transparent;
  border-style: none;
  border-bottom: 1px solid #ddd;
  padding: 20px 20px;
  justify-content: space-between;
}

.books-list-item-select {
  background-color: #e1e1e1;
}

.books-list-last-item {
  border-bottom-width: 0px;
}

.books-list-title {
  color: #7d6959;
  font-size: 20px;
  font-weight: bold;
}

.books-list-detail-header {
  color: #555;
  font-size: 14px;
  font-weight: normal;
}

.books-list-detail-value {
  color: #888;
  font-size: 16px;
  font-weight: bold;
}

.top-row {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  margin: 20px;
}

.search-input {
  overflow: hidden;
  border-radius: 15px;
}

.empty-state {
  color: #7d6959;
  font-size: 20px;
  font-weight: bold;
  text-align: center;
}

.cart-button {
  outline: none;
  display: flex;
  justify-content: space-around;
  align-items: center;
  min-width: 120px;
  margin-left: 10px;
  border-radius: 15px;
  overflow: hidden;
  text-align: center;
}
