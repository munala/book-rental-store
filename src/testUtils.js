import React from "react";
import {
  createHistory,
  createMemorySource,
  LocationProvider
} from "@reach/router";
import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";

export const dispatchMock = jest.fn();

export const customer = {
  id: 1,
  email: "munalaoliverngoitsi@gmail.com"
};

export const loading = false;
export const error = null;

export const books = [
  {
    id: 1,
    title: "Jurassic Park",
    author: "Michael Crichton",
    genre: "action"
  }
];

export const message = "message";

export const LoadingContext = React.createContext(loading);
export const ErrorContext = React.createContext(error);
export const CustomerContext = React.createContext(customer);
export const BooksContext = React.createContext(books);

export const renderWithRouter = (
  ui,
  { route = "/", history = createHistory(createMemorySource(route)) } = {}
) => {
  return {
    ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
    history
  };
};

export const LoadingProvider = ({ children }) => (
  <LoadingContext.Provider value={{ customer, loadingDispatch: dispatchMock }}>
    {children}
  </LoadingContext.Provider>
);

export const ErrorProvider = ({ children }) => (
  <ErrorContext.Provider value={{ errorDispatch: dispatchMock }}>
    {children}
  </ErrorContext.Provider>
);

export const CustomerProvider = ({ children }) => (
  <CustomerContext.Provider value={{ customerDispatch: dispatchMock }}>
    {children}
  </CustomerContext.Provider>
);

export const BooksProvider = ({ children }) => (
  <BooksContext.Provider value={{ booksDispatch: dispatchMock }}>
    {children}
  </BooksContext.Provider>
);

export const App = ({ children }) => (
  <LoadingProvider>
    <ErrorProvider>
      <CustomerProvider>
        <BooksProvider>{children}</BooksProvider>
      </CustomerProvider>
    </ErrorProvider>
  </LoadingProvider>
);

export const MockCustomerComponent = ({
  customer: mockCustomer,
  logout,
  onSubmit,
  loginMode,
  message: mockMessage,
  setMessage
}) => {
  return (
    <div>
      <button
        id="onSubmit"
        onClick={() => {
          onSubmit(customer);
        }}
      />
      <button
        id="logout"
        onClick={() => {
          logout(customer);
        }}
      />
      <button
        id="message"
        onClick={() => {
          setMessage(message);
        }}
      />
      {mockCustomer && <div id="customerId">{mockCustomer.id}</div>}
      <div id="loginMode">{loginMode}</div>
      <div id="message">{mockMessage}</div>
    </div>
  );
};

export const MockBooksComponent = ({ books: mockBooks, loadBooks, search }) => {
  return (
    <div>
      <button
        id="loadBooks"
        onClick={() => {
          loadBooks();
        }}
      />

      <button
        id="search"
        onClick={() => {
          search();
        }}
      />

      {books?.map(book => {
        return (
          <div id={book.id} key={book.id}>
            {book.id}
          </div>
        );
      })}
    </div>
  );
};
