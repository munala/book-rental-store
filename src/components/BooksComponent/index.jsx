import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import ListGroup from "react-bootstrap/ListGroup";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import Badge from "react-bootstrap/Badge";
import BooksCart from "../BooksCart/";
import "./styles.css";

const BooksComponent = ({ books, loadBooks, search }) => {
  const [searchTerm, setSearchTerm] = useState("");
  const [cart, setCart] = useState([]);
  const [showCart, setShowCart] = useState(false);

  useEffect(() => {
    loadBooks();
  }, []);

  useEffect(() => {
    if (!cart.length && showCart) {
      setShowCart(false);
    }
  }, [cart]);

  useEffect(() => {
    if (searchTerm) {
      search({
        title: searchTerm
      });
    } else {
      search({});
    }
  }, [searchTerm]);

  const checkInCart = bookId => {
    return cart.some(item => item.id === bookId);
  };

  const addToCart = book => {
    setCart(cart.concat(book));
  };

  const removeFromCart = book => {
    setCart(cart.filter(item => item.id !== book.id));
  };

  const clearCart = () => {
    setShowCart(false);
    setCart([]);
  };

  return (
    <div className="books-container">
      {showCart ? (
        <BooksCart
          cart={cart}
          clearCart={clearCart}
          removeFromCart={removeFromCart}
          hide={() => {
            setShowCart(false);
          }}
        />
      ) : (
        <>
          <div className="top-row">
            <Form.Control
              placeholder="start typing to search"
              value={searchTerm}
              aria-label="search"
              onChange={({ target: { value } }) => setSearchTerm(value)}
              className="search-input"
            />

            <Button
              disabled={!cart.length}
              variant="secondary cart-button"
              onClick={() => setShowCart(true)}
            >
              View Cart
              <Badge variant="primary">{cart.length || ""}</Badge>
              <span className="sr-only">unread messages</span>
            </Button>
          </div>

          <div className="books-list-container">
            <ListGroup as="ul" className="books-list">
              {books.map((book, index, list) => {
                const isInCart = checkInCart(book.id);
                const action = isInCart ? removeFromCart : addToCart;

                return (
                  <ListGroup.Item
                    key={book.id}
                    as="li"
                    className={`books-list-item ${
                      index === list.length - 1 ? "books-list-last-item" : ""
                    } ${isInCart ? " books-list-item-select" : ""}`}
                  >
                    <div>
                      <div className="books-list-title">{book.title}</div>

                      <div className="books-list-detail">
                        <i className="books-list-detail-header">by</i>&nbsp;
                        <span className="books-list-detail-value">
                          {book.author}
                        </span>
                      </div>

                      <div className="books-list-detail">
                        <i className="books-list-detail-header">under</i>&nbsp;
                        <span className="books-list-detail-value">
                          {book.genre}
                        </span>
                      </div>
                    </div>

                    <Button
                      variant={`${
                        isInCart ? "outline-" : ""
                      }primary center-align`}
                      onClick={() => action(book)}
                    >{`${isInCart ? "-" : "+"} ${
                      isInCart ? "Remove From" : "Add to"
                    } Cart`}</Button>
                  </ListGroup.Item>
                );
              })}
            </ListGroup>

            {!books.length && (
              <div className="empty-state">
                Nothing to show here. <br />
                Change or clear search term to display books.
              </div>
            )}
          </div>
        </>
      )}
    </div>
  );
};

BooksComponent.propTypes = {
  loadBooks: PropTypes.func.isRequired,
  search: PropTypes.func.isRequired,
  books: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      title: PropTypes.string,
      author: PropTypes.string,
      genre: PropTypes.genre
    }).isRequired
  )
};

export default BooksComponent;
