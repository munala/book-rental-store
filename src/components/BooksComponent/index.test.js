import React from "react";
import renderer from "react-test-renderer";
import BooksComponent from "./";
import { books } from "../../testUtils";

describe("BooksComponent tests", () => {
  test("should render", () => {
    const loadBooks = jest.fn();
    const search = jest.fn();

    const tree = renderer
      .create(
        <BooksComponent loadBooks={loadBooks} search={search} books={books} />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
