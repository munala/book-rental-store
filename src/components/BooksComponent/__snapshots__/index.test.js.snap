// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooksComponent tests should render 1`] = `
<div
  className="books-container"
>
  <div
    className="top-row"
  >
    <input
      aria-label="search"
      className="search-input form-control"
      onChange={[Function]}
      placeholder="start typing to search"
      value=""
    />
    <button
      className="btn btn-secondary cart-button"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      View Cart
      <span
        className="badge badge-primary"
      >
        
      </span>
      <span
        className="sr-only"
      >
        unread messages
      </span>
    </button>
  </div>
  <div
    className="books-list-container"
  >
    <ul
      className="books-list list-group"
      onKeyDown={[Function]}
    >
      <li
        className="books-list-item books-list-last-item  list-group-item"
        data-rb-event-key={null}
        disabled={false}
        onClick={[Function]}
      >
        <div>
          <div
            className="books-list-title"
          >
            Jurassic Park
          </div>
          <div
            className="books-list-detail"
          >
            <i
              className="books-list-detail-header"
            >
              by
            </i>
             
            <span
              className="books-list-detail-value"
            >
              Michael Crichton
            </span>
          </div>
          <div
            className="books-list-detail"
          >
            <i
              className="books-list-detail-header"
            >
              under
            </i>
             
            <span
              className="books-list-detail-value"
            >
              action
            </span>
          </div>
        </div>
        <button
          className="btn btn-primary center-align"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          + Add to Cart
        </button>
      </li>
    </ul>
  </div>
</div>
`;
